src/topology/category/Top/limits.lean

```diff
@@ -125,6 +126,452 @@ instance forget_preserves_colimits : preserves_colimits (forget : Top.{u} ⥤ Ty
     by exactI preserves_colimit_of_preserves_colimit_cocone
       (colimit_cocone_is_colimit F) (types.colimit_cocone_is_colimit (F ⋙ forget)) } }
 
+
+/-- The explicit fan of a family of topological spaces given by the pi type. -/
+@[simps X π_app]
+def pi_fan {ι : Type u} (α : ι → Top.{u}) : fan α :=
+@fan.mk _ _ _ α (Top.of (Π i, α i)) (λ i, ⟨λ f, f i, continuous_apply i⟩)
+
+/-- The constructed fan is indeed a limit -/
+def pi_fan_is_limit {ι : Type u} (α : ι → Top.{u}) : is_limit (pi_fan α) :=
+{ lift := λ S,
+  { to_fun := λ s i, S.π.app i s,
+    continuous_to_fun := by { apply continuous_pi, intro i, dsimp, continuity } },
+  fac' := by tidy,
+  uniq' := begin
+    intros S m h,
+    ext x i,
+    simp [← h i]
+  end }
+
+/--
+The product is homeomorphic to the product of the underlying spaces,
+equipped with the product topology.
+-/
+def pi_iso_pi {ι : Type u} (α : ι → Top.{u}) : ∏ α ≅ Top.of (Π i, α i) :=
+(limit.is_limit _).cone_point_unique_up_to_iso (pi_fan_is_limit α)
+
+@[simp, reassoc]
+lemma pi_iso_pi_inv_π {ι : Type u} (α : ι → Top) (i : ι) :
+  (pi_iso_pi α).inv ≫ pi.π α i = ⟨λ f, f i, continuous_apply i⟩ :=
+by simpa [pi_iso_pi]
+
+@[simp]
+lemma pi_iso_pi_inv_π_apply {ι : Type u} (α : ι → Top.{u}) (i : ι) (x : Π i, α i) :
+  (pi.π α i : _) ((pi_iso_pi α).inv x) = x i :=
+concrete_category.congr_hom (pi_iso_pi_inv_π α i) x
+
+@[simp]
+lemma pi_iso_pi_hom_apply {ι : Type u} (α : ι → Top.{u}) (i : ι) (x : ∏ α) :
+  (pi_iso_pi α).hom x i = (pi.π α i : _) x :=
+begin
+  have := pi_iso_pi_inv_π α i,
+  rw iso.inv_comp_eq at this,
+  exact concrete_category.congr_hom this x
+end
+
+/-- The explicit cofan of a family of topological spaces given by the sigma type. -/
+@[simps X ι_app]
+def sigma_cofan {ι : Type u} (α : ι → Top.{u}) : cofan α :=
+@cofan.mk _ _ _ α (Top.of (Σ i, α i)) (λ b, ⟨sigma.mk b⟩)
+
+/-- The constructed cofan is indeed a colimit -/
+def sigma_cofan_is_colimit {ι : Type u} (α : ι → Top.{u}) : is_colimit (sigma_cofan α) :=
+{ desc := λ S,
+  { to_fun := λ s, S.ι.app s.1 s.2,
+    continuous_to_fun := by { apply continuous_sigma, intro i, dsimp, continuity } },
```

Comment:
again, I think this can be handled by the `auto_param`.

Reply:
It seems that `continuity` could not solve it in this case: it cannot reduce structure projections.

Reply:
Oh, that's a shame.